cmake/Re2cBuildType.cmake

```diff
@@ -1,14 +1,76 @@
-set(_allowed_build_types Debug Release RelWithDebInfo MinSizeRel)
-
-# 1. The default value for CMAKE_BUILD_TYPE is an empty string.
-# 2. User can set CMAKE_BUILD_TYPE to any value at the cmake command line.
-#
-# Therefore, we check both cases, and make sure that we are dealing with
-# a known build type (if any).
-#
-# For more see discussion at https://github.com/skvadrik/re2c/pull/316
-if(CMAKE_BUILD_TYPE AND NOT CMAKE_BUILD_TYPE IN_LIST _allowed_build_types)
-    message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
+# CMAKE_BUILD_TYPE is not used by multi-configuration generators.
+# Thus, we need to check this first.
+get_property(isMultiConfig GLOBAL
+    PROPERTY GENERATOR_IS_MULTI_CONFIG)
+
+if(isMultiConfig) # Xcode, Visual Studio, ...
+    if (NOT WIN32)
+        if(NOT "Asan" IN_LIST CMAKE_CONFIGURATION_TYPES)
+            list(APPEND CMAKE_CONFIGURATION_TYPES Asan)
+        endif()
+        if(NOT "Ubsan" IN_LIST CMAKE_CONFIGURATION_TYPES)
+            list(APPEND CMAKE_CONFIGURATION_TYPES Asan)
+        endif()
+    endif()
+else() # Makefiles, Ninja, ...
+    if (NOT WIN32)
+        set(_allowed_build_types Asan Ubsan Debug Release RelWithDebInfo MinSizeRel)
+    else()
+        set(_allowed_build_types Debug Release RelWithDebInfo MinSizeRel)
+    endif()
+
+    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY
+        STRINGS "${_allowed_build_types}")
+
+    # 1. The default value for CMAKE_BUILD_TYPE is an empty string.
+    # 2. User can set CMAKE_BUILD_TYPE to any value at the cmake command line.
+    #
+    # Therefore, we check both cases, and make sure that we are dealing with
+    # a known build type (if any).
+    #
+    # For more see discussion at https://github.com/skvadrik/re2c/pull/316
+    if(CMAKE_BUILD_TYPE AND NOT CMAKE_BUILD_TYPE IN_LIST _allowed_build_types)
+        message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
+    endif()
+
+    unset(_allowed_build_types)
+endif()
+
+# AddressSanitizer (ASan) for Windows with MSVC is under experimental stage
```

Comment:
Minor thing, but some comments spell "ASan" and some "Asan", can you unify that? Either way is fine by me. And similarly with Ubsan.

Should the comment mention Ubsan as well, since it's also disabled on Windows?